src/main/scala/com/rbmhtechnology/calliope/scaladsl/TransactionalEventProducer.scala

```diff
@@ -203,68 +253,140 @@ private class GraphRunnerSupervisor(graph: ProducerGraph, onFailure: FailureHand
-        case Failure(err) => err
-        case Success(_) => new UnexpectedStreamCompletionException("Transactional producer stream stopped unexpectedly")
-      }
-      self ! RestartGraph(cause)
+    ft recover {
+      case _: GraphAbortedException => Done
+    } onComplete {
+      case Failure(err) => self ! RestartGraph(err)
+      case Success(_) => self ! GraphStopped
     }
-    queue
+    (queue, ks)
   }
 
-  private var commitQueue: SourceQueueWithComplete[Unit] = _
+  startWith(Stopped, Data())
+
+  when(Stopped) {
+    case Event(RunGraph, data) =>
+      log.info("Starting transactional producer stream.")
+      val (queue, ks) = runGraph()
+      goto(Running) using data.withQueue(queue).withKillSwitch(ks) replying GraphStarted
 
-  override def preStart(): Unit = {
-    self ! StartGraph
+    case Event(StopGraph, _) =>
+      stay replying GraphStopped
   }
 
-  override def postStop(): Unit = {
-    materializer.shutdown()
-    super.postStop()
+  when(Running) {
+    case Event(RunGraph, _) =>
+      stay replying GraphStarted
+
+    case Event(StopGraph, data) =>
+      log.info("Shutting down transactional producer stream.")
+      data.killSwitch.foreach(_.abort(GraphAbortedException()))
+      goto(Stopping) using data.withNotificationFor(sender())
+
+    case Event(RestartGraph(failure), data) =>
+      log.warning("Transactional producer stream was shutdown with cause: {}. Producer will be restarted", failure)
+      val (queue, ks) = runGraph()
+      stay using data.withQueue(queue).withKillSwitch(ks)
+
+    case Event(NotifyCommit, data) =>
+      data.commitQueue.foreach(_.offer(Unit))
+      stay
   }
 
-  private def initializing: Receive = {
-    case StartGraph =>
-      log.info("Starting transactional producer.")
-      commitQueue = runGraph()
+  when(Stopping) {
+    case Event(RunGraph, _) =>
+      stash()
+      stay
+
+    case Event(StopGraph, _) =>
+      stay replying GraphStopped
+
+    case Event(GraphStopped, data) =>
+      log.info("Transactional producer stream stopped.")
+      data.notifyOnStop.foreach(_ ! GraphStopped)
       unstashAll()
-      context.become(initialized)
+      goto(Stopped) using data.withoutNotifications()
+  }
 
-    case RestartGraph(_) =>
-      log.info("Restart of transactional producer requested while initialization in progress. Restart request is ignored.")
+  whenUnhandled {
+    case Event(NotifyCommit, _) =>
+      stay
 
-    case NotifyCommit =>
-      stash()
+    case Event(ev, _) =>
+      log.info(s"Received event $ev in state $stateName. Event is ignored.")
+      stay
   }
 
-  private def initialized: Receive = {
-    case NotifyCommit =>
-      commitQueue.offer(Unit)
+  initialize()
 
-    case RestartGraph(failure) =>
-      log.warning("Transactional producer was shutdown with cause: {}. Producer will be restarted", failure)
-      commitQueue = runGraph()
+  override def preRestart(reason: Throwable, message: Option[Any]): Unit = {
```

Comment:
you are right - the sender would get no notification of a successful start - we should only restart in case the stream was running.